### PR TITLE
Filter ANSI Escape Sequences

### LINKED ISSUE #62: Colors in gdb split
Hey, I saw that its possible to colorize the gdb output with a .gdbinit file like [this](https://gist.github.com/CocoaBeans/1879270) or [this](https://github.com/gdbinit/gdbinit) one. Using this with nvim-gdb, my source window stays empty. Is it possible to use such a config file with nvim-gdb?

### PR DESCRIPTION
**Pull request by mightymicha**
Hi here the requested PR. It does not fix #62 though, because the linked gdbinit files also change the output somehow, and the regex probably failes to act on these changes.

### CODE REVIEW

**Comment by mightymicha** [comment]
Such easy an easy .gdbinit file as:
`set prompt \033[31m(gdb) \033[0m`
fails after start.

**Comment by sakhnik** [comment]
The change itself passes the tests!

File: rplugin/python3/gdb/scm.py [hunk 1]
```diff
@@ -49,6 +50,9 @@ def queryB(self, match):
     def _search(self):
         # If there is a matcher matching the line, call its handler.
         for matcher, func in self.state:
+            # Filter ANSI Escape Sequence
+            reg = r'(\x9B|\x1B\[)[0-?]*[ -\/]*[@-~]'
```
Comment (sakhnik) [inline rplugin/python3/gdb/scm.py]: I believe there could be a better place for filtering out. The `self.buffer` is an accumulation buffer, and thus, your substitution will be doing repetitive searches in it. It'd be better to hook up before the method `_search` is called. Somewhere closer to the data source.

**Comment by sakhnik** [comment]
Would you be able to log the filtered data to a file and compare the stream with the pristine GDB output?

File: rplugin/python3/gdb/scm.py (hunk 1 above)
Comment (sakhnik) [reply rplugin/python3/gdb/scm.py]: On the other hand, the stream doesn't guarantee that an escape sequence is received atomically. So I recommend moving the substitution just of the matching loop.
Does it substitute all the occurrences?

**Comment by mightymicha** [comment]
> Would you be able to log the filtered data to a file and compare the stream with the pristine GDB output?

[Here](https://gist.github.com/mightymicha/dfb1841e843c7f148ac3aad4320015dd) is the logged output with script of some example code, the last one is the ANSI cleaned gdbinit version. But it is a bit messy. 

[Here](https://gist.github.com/mightymicha/6d718b9da348b57dc6f2011fb603eeb1) is the output when logging directly after applying the regex (moved outside of the match function). The input field in the gdb window is unresponsive after printing:

`server info breakpoints
No breakpoints or watchpoints.`

File: rplugin/python3/gdb/scm.py (hunk 1 above)
Comment (mightymicha) [reply rplugin/python3/gdb/scm.py]: > Does it substitute all the occurrences?

Found the regex [here](https://stackoverflow.com/questions/14693701/how-can-i-remove-the-ansi-escape-sequences-from-a-string-in-python)

**Comment by sakhnik** [comment]
Ups, responded in the issue: https://github.com/sakhnik/nvim-gdb/issues/62#issuecomment-474575586
To make it work, we'll need to also modify the class `_StringMatcher` to skip over escape sequences. This class is used by the `StreamFilter` to determine when to stop passing data through to the terminal, and when to stop doing so. The prompt is the terminal symbol.
Alternatively, we could you GDB's command `show prompt` and use its output both in the SCM and in the `StreamFilter`.

**Comment by sakhnik** [comment]
I've opened a branch `prompt` with a couple of changes to the gdb proxy. It does seem reasonable to query the current prompt sometimes and adapt both the proxy and the SCM accordingly. The plugin will request the prompt once on launch and every time the breakpoints are queried.
Please feel free to experiment. Probably, your change should work in this branch already assuming the prompt stays "(gdb) ".
I'll continue working on the change soon.

**Comment by sakhnik** [comment]
I've got new ideas that I'm exploring now:
- Query the prompt and adopt both the proxy and the scm to it (talked about it above)
- Forge the prompt to contain marker symbols like ^Z (pros: rigid SCM and proxy, cons: require a user to abstain changing the prompt in run time)
- Use the Python interpreter inside GDB to support the side channel instead of proxying and parsing the GDB CLI output.

The last one is actually the most attractive, and that's how LLDB is harnessed. But I still can't figure out how to efficiently list breakpoints using Python in GDB. Combining with #53, it'd be possible to forget about prompt forever! So still experimenting and designing in my free time.

**Comment by sakhnik** [comment]
Not needed since I used `gdb.prompt_hook`

**Pull request closed without merging**